design/service-apis-implementation.md

```diff
@@ -0,0 +1,342 @@
+There should never be annotations required for functionality on Service-APIs objects.
+
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support, covering the v1alpha1 version of the service-apis.
+- Layer 7 support only, which means HTTPRoutes, TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
+This will be a configuration file parameter.
+When a Gateway is not supplied, Contour will not do Service-APIs processing, or in other words, the existing `--experimental-service-apis` flag will be obsoleted and removed).
+Inside Contour will merge Listeners within its Gateway. (See the detailed design for the exact rules Contour will use for this.)
+
+### Interrelated watches
+
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with its named Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
```

Comment:
Some aspects of this are not entirely clear to me, e.g. what if there's a service-apis Listener with a wildcard domain name that overlaps with an HTTPProxy with a specific domain name, or what if some other service-apis Listener settings conflict with Contour's listener settings -- but I'm also not sure it's that important right now, and we can figure it out during implementation.

Reply:
I think the wildcard domain name thing will be sorted out by implementing wildcard domains everywhere - we should do that before starting with this, since we need to do it for Ingress v1 anyway. I think the limited form of wildcard FQDN defined in the service-apis is probably implementable in HTTPProxy as well.

I think the LIstener settings details should be sorted out during implementation, but I think that the service-apis spec should probably override the Contour config one, since it's more specific (I think). I don't think it matters really, as long as we document it.